#! /bin/sh /usr/share/dpatch/dpatch-run
## 05_xenrt by  <mnencia@debian.org>
##
## DP: This patch will allow nvidia to be compile and run on Xen or
## DP: Realtime Preemption enabled kernels.

@DPATCH@
Index: usr/src/nv/nv-linux.h
===================================================================
--- usr/src/nv/nv-linux.h.orig	2007-03-26 15:58:49.000000000 +0200
+++ usr/src/nv/nv-linux.h	2007-03-26 15:59:11.000000000 +0200
@@ -244,7 +244,7 @@
  * tiny, and the kernel panics when it is exhausted. try to warn the user that
  * they need to boost the size of their pool.
  */
-#if defined(CONFIG_SWIOTLB) && !defined(GFP_DMA32)
+#if defined(CONFIG_SWIOTLB) && !defined(GFP_DMA32) && !defined(CONFIG_XEN)
 #define NV_SWIOTLB 1
 #endif
 
@@ -776,7 +776,10 @@
 #define NV_VM_INSERT_PAGE(vma, addr, page) \
     vm_insert_page(vma, addr, page)
 #endif
-#if defined(NV_REMAP_PFN_RANGE_PRESENT)
+#if defined(CONFIG_XEN)
+#define NV_REMAP_PAGE_RANGE(from, offset, x...) \
+    io_remap_pfn_range(vma, from, ((offset) >> PAGE_SHIFT), x)
+#elif defined(NV_REMAP_PFN_RANGE_PRESENT)
 #define NV_REMAP_PAGE_RANGE(from, offset, x...) \
     remap_pfn_range(vma, from, ((offset) >> PAGE_SHIFT), x)
 #elif defined(NV_REMAP_PAGE_RANGE_5_PRESENT)
@@ -788,6 +791,9 @@
 #define NV_REMAP_PAGE_RANGE(x...) remap_page_range(x)
 #endif
 
+#if !defined(CONFIG_XEN)
+#define phys_to_machine(x) x
+#endif
 
 #define NV_PGD_OFFSET(address, kernel, mm)              \
    ({                                                   \
Index: usr/src/nv/nv-vm.c
===================================================================
--- usr/src/nv/nv-vm.c.orig	2007-03-26 15:58:49.000000000 +0200
+++ usr/src/nv/nv-vm.c	2007-03-26 15:59:11.000000000 +0200
@@ -352,6 +352,9 @@
 
 static void nv_flush_caches(void)
 {
+#if defined(CONFIG_PREEMPT_RT)
+    if(!nv_pat_enabled) return;
+#endif
 #if defined(KERNEL_2_4)
     // for 2.4 kernels, just automatically flush the caches and invalidate tlbs
     nv_execute_on_all_cpus(cache_flush, NULL);
@@ -502,7 +505,7 @@
         page_ptr->phys_addr = phys_addr;
         page_ptr->page_count = NV_GET_PAGE_COUNT(page_ptr);
         page_ptr->virt_addr = virt_addr;
-        page_ptr->dma_addr = page_ptr->phys_addr;
+        page_ptr->dma_addr = phys_to_machine(page_ptr->phys_addr);
 
         /* lock the page for dma purposes */
         nv_lock_page(page_ptr);
Index: usr/src/nv/nv.c
===================================================================
--- usr/src/nv/nv.c.orig	2007-03-26 15:58:49.000000000 +0200
+++ usr/src/nv/nv.c	2007-03-26 15:59:11.000000000 +0200
@@ -42,8 +42,26 @@
 
 int nv_pat_enabled = 0;
 
+/*
+ * disable PAT support if XEN or PREEMPT_RT is configured in kernel
+ */
+
+#if defined(CONFIG_XEN) || defined(CONFIG_PREEMPT_RT)
+static int nv_disable_pat = 1;
+#else
 static int nv_disable_pat = 0;
+#endif
+
+/*
+ * you can re-enable PAT support for PREEMPT_RT when applying
+ * "nv_disable_pat=0" as kernel parameter for the sake of slightly
+ * better 3D performance but at the expense of higher latencies.
+ * if XEN is configured, then PAT support can't be enabled!
+ */
+
+#if !defined(CONFIG_XEN)
 NV_MODULE_PARAMETER(nv_disable_pat);
+#endif
 
 #if defined(NVCPU_X86) || defined(NVCPU_X86_64)
 NvU64 __nv_supported_pte_mask = ~_PAGE_NX;
Index: usr/src/nv/os-agp.c
===================================================================
--- usr/src/nv/os-agp.c.orig	2007-03-26 15:58:49.000000000 +0200
+++ usr/src/nv/os-agp.c	2007-03-26 15:59:11.000000000 +0200
@@ -286,7 +286,7 @@
 
          page_ptr->phys_addr = (ptr->memory[i] & PAGE_MASK);
          page_ptr->virt_addr = (unsigned long) __va(page_ptr->phys_addr);
-         page_ptr->dma_addr  = page_ptr->phys_addr;
+         page_ptr->dma_addr  = phys_to_machine(page_ptr->phys_addr);
     }
 
     return RM_OK;
Index: usr/src/nv/os-interface.c
===================================================================
--- usr/src/nv/os-interface.c.orig	2007-03-26 15:58:49.000000000 +0200
+++ usr/src/nv/os-interface.c	2007-03-26 15:59:11.000000000 +0200
@@ -533,6 +533,7 @@
     MicroSeconds = MilliSeconds * 1000;
     tm_end.tv_usec = MicroSeconds;
     tm_end.tv_sec = 0;
+#if !defined(CONFIG_XEN)
     NV_TIMERADD(&tm_aux, &tm_end, &tm_end);
 
     /* do we have a full jiffie to wait? */
@@ -570,6 +571,7 @@
                 MicroSeconds = 0;
         } while ((jiffies = NV_USECS_TO_JIFFIES(MicroSeconds)) != 0);
     }
+#endif
 
     if (MicroSeconds > 1000)
     {
Index: usr/src/nv/conftest.sh
===================================================================
--- usr/src/nv/conftest.sh.orig	2007-03-26 15:58:49.000000000 +0200
+++ usr/src/nv/conftest.sh	2007-03-26 15:59:11.000000000 +0200
@@ -901,7 +901,8 @@
         # Check if the target kernel is a Xen kernel. If so, then exit, since
         # the driver doesn't currently work with Xen.
         #
-        RET=1
+        ##MNENCIA## Switched to 0 because we have the xenrt patch :-)
+        RET=0
         VERBOSE=$6
         FILE="linux/autoconf.h"
 
